lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1627,18 +1677,21 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
       return true;
     }
 
-    static bool implicitWeakSelfReferenceIsValid(const DeclRefExpr *DRE) {
-      ASTContext &Ctx = DRE->getDecl()->getASTContext();
+    static bool
+    implicitWeakSelfReferenceIsValid(const DeclRefExpr *DRE,
+                                     const AbstractClosureExpr *inClosure) {
+      auto selfDecl = valueDeclForDRE(DRE, inClosure);
```

Comment:
Then, when we call the `getCapturedSelfDecl` here, we should leave a big comment detailing the warning suppression heuristic, and how it's okay to suppress the warning in Swift 5 because although the formal declaration for implicit `self` will change in Swift 6, there are no observable semantic differences in the resulting code. Writing out the code example that you left in your earlier comment might even be helpful!